sysbuild: Image ordering

57300 opened the pull request:
Set the configuration or flashing order for images:
```
sysbuild_add_dependencies(CONFIGURE my_sample sample_a sample_b)
sysbuild_add_dependencies(FLASH     my_sample sample_c sample_d)
```

Please review the doc patch for more information.

Fixes #53650

57300 committed:
sysbuild: Remove IMAGES variable

This variable was originally provided for two indended purposes:

  * Let users manually add a new image in a desired order in the list.
  * Let users set build-only images, which are excluded from the list.

Given the recent additions of the `sysbuild_add_dependencies()` function
and the `BUILD_ONLY` parameter, `IMAGES` should be considered obsolete.

Furthermore, the list of images added to sysbuild should be updated
automatically when calling `ExternalZephyrProject_Add()`. This is now
possible by using a GLOBAL property internal to sysbuild.

With that, the `IMAGES` variable can be removed. Its existing usage for
image ordering is replaced with `sysbuild_add_dependencies()` treewide.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>

57300 committed:
sysbuild: Support relative ordering of images

Fixes #53650

The existing solution for image ordering involves the `IMAGES` variable,
which sysbuild originally provided to let users manually add a new image
in a desired order. This isn't very flexible for the following reasons:

* The order in which `IMAGES` is updated across multiple modules and
  `sysbuild.cmake` files is not well defined.
* Having a single variable means that the same order is used for both
  configuration and flashing. Usually, there is no reason for the
  flashing order to be the same as the configuration order.

Introduce the `sysbuild_add_dependencies()` function for more fine-tuned
ordering of images. It makes one image depend on other images in either
configuration or flashing order. Its usage is similar to the standard
CMake function `add_dependencies()`, but with an extra parameter to
distinguish between two types of dependencies:

 sysbuild_add_dependencies(CONFIGURE my_sample sample_a sample_b)
 sysbuild_add_dependencies(FLASH     my_sample sample_c sample_d)

CONFIGURE dependencies determine the order in which sysbuild configures
(runs CMake for) the individual images. This is useful if there is some
information from one application's build which needs to be available to
another application.

FLASH dependencies control the sequence of images used by `west flash`.
This could be used if a specific flashing order is required by an SoC,
a runner, or something else. Note that these dependencies are not valid
for images specified as `BUILD_ONLY`.

The internal `sysbuild_images_order()` function is responsible for
assembling two sorted lists of images based on the added dependencies,
with the help of `topological_sort()`.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>

57300 committed:
doc: sysbuild: Add documentation for image ordering

Add a brief subsection to explain `sysbuild_add_dependencies()`.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>

nordicjm commented:
Just gave it a quick try and seems to have a problem if you order by images that are not present, i.e. no error is raised.
E.g. putting this in `hello_world/sysbuild.cmake`:
```
sysbuild_add_dependencies(IMAGE CONFIGURE hello_world loam)
sysbuild_add_dependencies(IMAGE FLASH     hello_world loam)
```
Configures and flashes fine like there is no error, but an error should be raised because there is no "loam" image.

nordicjm commented on doc/build/sysbuild/index.rst:
```diff
@@ -619,6 +619,51 @@ configuration files for ``my_sample`` will be ignored.
+
+.. tip::
+
+   Unlike with ``add_dependencies()``, you are allowed to use
+   ``sysbuild_add_dependencies()`` with non-existent images. This makes it
+   possible to define relationships between various images in the build system,
+   regardless of whether they are present in the current multi-image build.
```
Just saw this, I do not like this idea, why would you want to sort on images that are not present? They shouldn't be included in a build.

57300 replied:
I intended this initially to simplify adding dependencies for a conditionally included image, thus avoiding a lot of this:
```
if mcuboot exists:
  mcuboot depends on my_sample
endif
```

Similarly for groups:
```
if mcuboot exists:
  add mcuboot to my_group
endif
```

But on second thought, maybe these cases won't even be as common as I first imagined. I'll consider raising errors, just like the standard `add_dependencies()`.

57300 replied:
Actually, I just remembered that since the processing order of CMake files is not well defined, there might not even be a good way to check `if <image> exists`.

nordicjm replied:
After all images have been configured should be OK.

57300 replied:
At this point, I expect all groups and dependencies to be added as well. They are resolved before the PRE_CMAKE hooks.

57300 replied:
> Those will be processed in the order they are reported by `west list` (or `ZEPHYR_MODULES` when `west` is not used)

That's neat! Based on this fact, I added another change to make the order of including image-specific `sysbuild.cmake` well-defined. With that, I can comfortably address the lack of errors from `sysbuild_add_dependencies()`.

nordicjm commented on doc/build/sysbuild/index.rst:
```diff
@@ -648,3 +693,285 @@ each image that is part of a project. Alternatively, there are
+
+An immutable group can only be defined once. In this case, since ``my_group``
+already exists, this would produce an error, so you could use a different name
+to ensure that your group has a consistent meaning in the build system.
```
Should probably add a note with naming conventions for groups e.g. no spaces, semi-colons, commas, etc.

57300 replied:
Maybe. I suppose a note should be added somewhere for image names too.

nordicjm commented on doc/build/sysbuild/index.rst:
```diff
@@ -648,3 +693,285 @@ each image that is part of a project. Alternatively, there are
+      ┌───────────────────────────── unique to group_b
+      │                     ┌─────── unique to group_a
+     i_7        i_1        i_4
+   { i_8 } -> { i_2 } -> { i_5 }
```
A neat addition for the future would be a ninja command that prints out a graphic like this for a configuration showing the flash/configuration order

57300 replied:
Definitely. So far, I only added a partial printout for when you get a cyclic dependency error.

nordicjm commented on doc/build/sysbuild/index.rst:
```diff
@@ -648,3 +693,285 @@ each image that is part of a project. Alternatively, there are
+
+.. tip::
+
+   You are allowed to use ``sysbuild_add_dependencies()`` with non-existent
+   groups, just as well as with non-existent images.
```
And why?

tejlmand replied:
> For groups, I still don't want to raise an error. 

Raising errors when something is not defined is generally a good things.
It catches a lot of small mistakes early and avoid lot of wasted debugging time when something doesn't work as expected later during build or flash.

For example:
```
sysbuild_add_dependencies(GROUP CONFIGURE botloader ...) # <---- whoops, should have been bootloader, but not discovered cause non-existing groups are allowed.
```

57300 replied:
I understand, but raising an error would also clash with images/groups and their dependencies being defined out of order. See: https://github.com/zephyrproject-rtos/zephyr/pull/57884#discussion_r1193611885

tejlmand replied:
> but raising an error would also clash with images/groups and their dependencies being defined out of order.

dependencies should be defined after images / (groups) has been defined.
So that adding dependencies on something not defined fails immediately where the `sysbuild_add_dependencies()` is called. Anything else will make it much harder to debug, either because they are silently ignored, or because it's being warned / error at completely different location or at build time.

Such errors causes confusion as users must suddenly try to search for the location where the dependency is defined.

57300 replied:
> dependencies should be defined after images / (groups) has been defined.

Okay, then one option is to allow adding dependencies in `PRE_CMAKE`, as you suggested here: https://github.com/zephyrproject-rtos/zephyr/pull/57884#discussion_r1209993063

> So that adding dependencies on something not defined fails immediately

I think people might expect this function to behave the same way as the standard `add_dependencies(<target> <dependencies>)`. It only fails immediately if `<target>` doesn't exist, while `<dependencies>` are not checked until the end of the configuration stage.

So, that's what I would probably add for sysbuild, with the final checks being deferred to `sysbuild_images_sort()`.

57300 replied:
> adding dependencies on something not defined fails immediately where the `sysbuild_add_dependencies()` is called.

Addressed.

> a back-reference in case of a failure to where `sysbuild_add_dependencies()` was called would be great, but hard to achieve wrt. the line number.

Yeah, I wish I could add such back-references with correct filenames and line numbers. I tried out a few ways, but I couldn't get it right.

nordicjm reviewed (commented):
Just looked at docs thus far

tejlmand commented on share/sysbuild/cmake/modules/sysbuild_extensions.cmake:
```diff
@@ -353,6 +363,93 @@ function(ExternalZephyrProject_Cmake)
 endfunction()
 
+# Usage:
+#   ExternalZephyrProject_Group(<name>
```
Not sure about the `_Group` name here.

The `ExternalZephyrProject_Group()` seems to more handle properties and not be doing an explicit step as described:
https://cmake.org/cmake/help/latest/module/ExternalProject.html#explicit-step-management

That's a difference to the `ExternalZephyrProject_CMake()` function which actually carries out a dedicate step, which is the CMake configure step, but with the slight difference that the step is executed at sysbuild CMake configure time and not sysbuild build time (which would be the normal behavior of an external step).

tejlmand replied:
also see this comment: https://github.com/zephyrproject-rtos/zephyr/pull/57884#discussion_r1201620427

I believe groups should be independent of `ExternalZephyrProject` cause we want to allow handling of non-Zephyr projects.
As long as such project can provide / generate a runners we can still use west to flash them.

This indicates that the group feature should be independent of the `ExternalZephyrProject` feature.

57300 replied:
Continued: https://github.com/zephyrproject-rtos/zephyr/pull/58174#discussion_r1201660043

tejlmand commented on share/sysbuild/cmake/modules/sysbuild_extensions.cmake:
```diff
@@ -353,6 +363,93 @@ function(ExternalZephyrProject_Cmake)
+# EXCLUDE <images>:      Remove a set of images from the group.
+# INCLUDE_FROM <groups>: INCLUDE all images belonging to each group in the list.
+# EXCLUDE_FROM <groups>: EXCLUDE all images belonging to each group in the list.
+# IMMUTABLE:             Disallow extending the definition of this group.
```
At current stage I don't see any reasons why sysbuild should support this.

It will just result in users having to look into the sysbuild CMake code to understand how to manipulate the right properties, so atm it seems just providing false impression that a group is locked.

57300 replied:
Continued: https://github.com/zephyrproject-rtos/zephyr/pull/58174#discussion_r1201662733

tejlmand commented on share/sysbuild/cmake/modules/sysbuild_extensions.cmake:
```diff
@@ -353,6 +363,93 @@ function(ExternalZephyrProject_Cmake)
+  set_property(TARGET ${target_name} PROPERTY IMMUTABLE "${ZGROUP_IMMUTABLE}")
+  foreach(arg ${multi_args})
+    if(DEFINED ZGROUP_${arg})
+      set_property(TARGET ${target_name} APPEND PROPERTY ${arg} "${ZGROUP_${arg}}")
```
it just seems more overhead to keep track of both included and excluded.

57300 replied:
Continued: https://github.com/zephyrproject-rtos/zephyr/pull/58174#discussion_r1201664543

tejlmand commented on share/sysbuild/cmake/modules/sysbuild_extensions.cmake:
```diff
@@ -353,6 +363,93 @@ function(ExternalZephyrProject_Cmake)
+#                               [IMMUTABLE]
+#   )
+#
+# This function defines groups of external Zephyr projects. Images added using
+# `ExternalZephyrProject_Add()` may inherit common properties of the group(s)
+# they are assigned to.
```
> may inherit common properties of the group(s) they are assigned to.

this sound a bit fluffy.
Do they inherit properties, and if so, which properties a inherited ?

This starts to sound a bit like CMake's [Transitive Usage Requirements](https://cmake.org/cmake/help/latest/manual/cmake-buildsystem.7.html#target-usage-requirements), and if that is the case, then we should consider if those key words would be better to use in this context.

tejlmand reviewed (changes requested):
Some initial comments.

If possible, I would prefer if the whole group feature could be made as an independent PR, because I think some proposals in that part of the PR could result in some more discussion.

tejlmand commented on share/sysbuild/cmake/modules/sysbuild_extensions.cmake:
```diff
@@ -134,6 +134,8 @@ function(ExternalZephyrProject_Add)
     )
   endif()
 
+  set_property(GLOBAL APPEND PROPERTY sysbuild_images ${ZBUILD_APPLICATION})
```
This should not be added here.

The existing code at those places:
https://github.com/zephyrproject-rtos/zephyr/blob/7d898edfc5c2e8dfa4bc158c5f74614f15dd833b/share/sysbuild/CMakeLists.txt#L88
https://github.com/zephyrproject-rtos/zephyr/blob/7d898edfc5c2e8dfa4bc158c5f74614f15dd833b/share/sysbuild/bootloader/CMakeLists.txt#L13

are exactly done outside of `ExternalZephyrProject_Add()` in order to keep building of Zephyr based projects independent of flashing.
A use-case is where a sysbuild module / extension want to build a Zephyr based project but merge the output of such project into another image.

Also, not all images may be Zephyr project. There can be ather forms of external projects, which can be added to sysbuild image list. Of course flashing of such images requires that they generate a runners.yml file, but such file is quite simple.
Such a use-case could be porting of the espressif bootloader to sysbuild, ref:
https://github.com/zephyrproject-rtos/zephyr/blob/7d898edfc5c2e8dfa4bc158c5f74614f15dd833b/soc/xtensa/esp32s3/CMakeLists.txt#L29-L30

So please keep the list of sysbuild images to flash independent of the `ExternalZephyrProject_Add()`.

Of course this comment naturally also impact the later GROUP argument proposal, cause if `ExternalZephyrProject_Add()` is decoupled from sysbuild image list, then it should also be decoupled from sysbuild image groups.

57300 replied:
> The existing code at those places (...) are exactly done outside of `ExternalZephyrProject_Add()` in order to keep building of Zephyr based projects independent of flashing. A use-case is where a sysbuild module / extension want to build a Zephyr based project but merge the output of such project into another image (instead of flashing said Zephyr image directly).

You mean to support building an image without flashing it? I was not aware of this and I'm pretty sure this is undocumented. I would propose specifying that as an argument to `ExternalZephyrProject_Add()` as well, so that the `IMAGES` variable can be removed.

> Also, not all images may be Zephyr project. There can be other forms of external projects which can be added to sysbuild image list.

My impression was that non-Zephyr projects were not really closely tied to sysbuild. Here's the documentation:

https://github.com/zephyrproject-rtos/zephyr/blob/2f718dd369e41a9ae50fb613ac29d8ce192da440/doc/build/sysbuild/index.rst?plain=1#L622-L635

I wouldn't presume it necessary to populate `IMAGES` when using standard `ExternalProject_Add()`, because I could call it with `BUILD_ALWAYS TRUE` to ensure that it builds. For Zephyr projects, sysbuild will ignore them unless I append to `IMAGES`. Also, I'm not sure if there are any side effects of calling `ExternalZephyrProject_Cmake()` on a target not added using `ExternalZephyrProject_Add()`. All of this should probably be documented.

By the way, is there a reason why the espressif bootloader won't be able to use `ExternalZephyrProject_Add()`?

tejlmand replied:
> You mean to support building an image without flashing it?

Correct.

> I was not aware of this and I'm pretty sure this is undocumented. 

It's been discussed as part of a potential wish for systems to be able to build Zephyr samples, but decide to merge to single / fewer combined images instead of flashing each image independently. That would currently be a very advanced use, but is the reason why building and flashing is not intermixed, as that could make it harder to split / make flashing more independent of building later. 
Also, it's not directly described that `ExternalZephyrProject_Add()` will also do flashing: 
https://github.com/zephyrproject-rtos/zephyr/blob/7e23e73bc186da4192a75f367e8d4f53441016de/share/sysbuild/cmake/modules/sysbuild_extensions.cmake#L112-L113

But it's a very valid point that we should improve the doc details on this.
Especially after the split of adding the project and running CMake in this commit: https://github.com/zephyrproject-rtos/zephyr/commit/631fa636100b9605dfbc9bfb1968fdfae656ddd4 it is neccesary to ensure CMake is still executed, which is also done based on the `IMAGES` variable.

> I would propose specifying that as an argument to `ExternalZephyrProject_Add()` as well, so that the `IMAGES` variable can be removed.

Need some careful consideration. Adding an argument to `ExternalZephyrProject_Add()` would indicate it's the caller of the function who decides this, where in some cases it could actually be a downstream project who would like to control this.
It's a very valid discussion you're opening, and perhaps it's time to do some more consideration on this topic.

tejlmand replied:
> By the way, is there a reason why the espressif bootloader won't be able to use ExternalZephyrProject_Add()?

The fact that the espressif bootloader is not a Zephyr project 😉

tejlmand replied:
> Also, I'm not sure if there are any side effects of calling `ExternalZephyrProject_Cmake()` on a target not added using `ExternalZephyrProject_Add()`.

Handled here:
https://github.com/zephyrproject-rtos/zephyr/blob/215f180296c9148a468eb294090e0209629cc81f/share/sysbuild/cmake/modules/sysbuild_extensions.cmake#L298-L302

nordicjm replied:
> The fact that the espressif bootloader is not a Zephyr project wink

I don't believe that this is true anymore, I think it has been switched to mcuboot

tejlmand replied:
> I don't believe that this is true anymore, I think it has been switched to mcuboot

both MCUboot and Espressif IDF bootloaders are supported as 2nd stage bootloader for esp32.

And the support of MCUboot on esp32 still doesn't make the IDF bootloader a Zephyr project 😉

57300 commented:
> If possible, I would prefer if the whole group feature could be made as an independent PR, because I think some proposals in that part of the PR could result in some more discussion.

No problem. I made a separate PR for the grouping feature and will address relevant comments there: https://github.com/zephyrproject-rtos/zephyr/pull/58174

It's a draft, so please add yourself as a reviewer if you're interested.

57300 commented on share/sysbuild/cmake/modules/sysbuild_extensions.cmake:
```diff
@@ -451,3 +453,120 @@ function(sysbuild_cache_set)
+  endif()
+
+  set(property_name sysbuild_deps_${dependency_type}_${image})
+  set_property(GLOBAL APPEND PROPERTY ${property_name} ${ARGN})
```
I used global properties to support using `sysbuild_add_dependencies()` on images that haven't been added yet, in which case the target wouldn't exist.

See this related comment chain: https://github.com/zephyrproject-rtos/zephyr/pull/57884#discussion_r1193611885 I would appreciate your input there.

tejlmand replied:
> I used global properties to support using sysbuild_add_dependencies() on images that haven't been added yet, in which case the target wouldn't exist.

Not a good enough reason, cause it also make the rest of the code more complex.
If dependencies are added on targets you can pass the targets directly to your topology search and each target itself will know it's dependencies.

This mean that all code related to loop your global properties and creation of edges are can simply be removed, such as:
https://github.com/zephyrproject-rtos/zephyr/blob/5535dce7e745430f500a8849d40b718ad0821449/share/sysbuild/cmake/modules/sysbuild_extensions.cmake#L551-L566

https://github.com/zephyrproject-rtos/zephyr/blob/5535dce7e745430f500a8849d40b718ad0821449/share/sysbuild/cmake/modules/sysbuild_extensions.cmake#L491-L500

tejlmand replied:
> This mean that all code related to loop your global properties and creation of edges are can simply be removed, such as:

For example, the following function will work directly on target, and thus all other surrounding code can be eliminated.

Note: ported from here: https://github.com/zephyrproject-rtos/zephyr/blob/7457bcf0a2b3d3523b8b3ea971771337f3751f1f/scripts/zephyr_module.py#L603

```
function(topological_sort)
  cmake_parse_arguments(TS "" "RESULT;PROPERTY_NAME" "TARGETS" ${ARGN})
  
  set(dep_targets)
  set(start_targets)
  set(sorted_targets)
  
  foreach(target ${TS_TARGETS})
    get_target_property(${target}_dependencies ${target} ${TS_PROPERTY_NAME})
  
    if(${target}_dependencies)
      list(APPEND dep_targets ${target})
    else()
      list(APPEND start_targets ${target})
    endif()
  endforeach()
  
  while(TRUE)
    list(POP_FRONT start_targets node)
    list(APPEND sorted_targets ${node})
    set(to_remove)
    foreach(target ${dep_targets})
      if("${node}" IN_LIST ${target}_dependencies)
        list(REMOVE_ITEM ${target}_dependencies ${node})
        if(NOT ${target}_dependencies)
          list(APPEND start_targets ${target})
          list(APPEND to_remove ${target})
        endif()
      endif()
    endforeach()
  
    foreach(target ${to_remove})
      list(REMOVE_ITEM dep_targets ${target})
    endforeach()
    if(NOT start_targets)
      break()
    endif()
  endwhile()
  
  if(dep_targets)
    foreach(target ${dep_targets})
      get_target_property(deps ${target} ${TS_PROPERTY_NAME})
      list(JOIN deps " " deps)
      list(APPEND dep_string "${target} depends on: ${deps}")
    endforeach()
    list(JOIN dep_string "\n" dep_string)
    message(FATAL_ERROR "Unmet or cyclic dependencies in modules:\n${dep_string}")
  endif()
  
  set(${TS_RESULT} "${sorted_targets}" PARENT_SCOPE)
endfunction()
```

Function can easily be tested, like this:
```
add_custom_target(A)
add_custom_target(B)
add_custom_target(C)
add_custom_target(D)

set(IMAGES A B C D)
message("Unsorted images: ${IMAGES}")

set_property(TARGET A APPEND PROPERTY CUSTOM_DEPENDS C)
set_property(TARGET B APPEND PROPERTY CUSTOM_DEPENDS A C)
set_property(TARGET C APPEND PROPERTY CUSTOM_DEPENDS D)

topological_sort(TARGETS ${IMAGES} PROPERTY_NAME CUSTOM_DEPENDS RESULT IMAGES)
message("Sorted images  : ${IMAGES}")
message("==========================\n")

# Add cyclic dep.
set_property(TARGET C APPEND PROPERTY CUSTOM_DEPENDS A)
topological_sort(TARGETS ${IMAGES} PROPERTY_NAME CUSTOM_DEPENDS RESULT IMAGES)
message("Resorted images  : ${IMAGES}")
```
Result:
```
Unsorted images: A;B;C;D                                                                                                                                                                                
Sorted images  : D;C;A;B                                                                                                                                                                                
==========================                                                                                                                                                                              
```

And cyclic errors:
```
CMake Error at CMakeLists.txt:54 (message):                                                                                                                                                             
  Unmet or cyclic dependencies in modules:                                                                                                                                                              
                                                                                                                                                                                                        
  C depends on: A                                                                                                                                                                                       
                                                                                                                                                                                                        
  A depends on: C                                                                                                                                                                                       
                                                                                                                                                                                                        
  B depends on: A                                                                                                                                                                                       
Call Stack (most recent call first):                                                                                                                                                                    
  CMakeLists.txt:79 (topological_sort)              
```

The topological sort is very similar to your current proposal, and feel free to rework that one into use targets (or images) instead of using above proposal.

The main point is to avoid all the additional overhead / maintenance of extra global properties, processing of those properties into additional `,`-separated strings and then pass this on to the sort function when everything can be done directly using target properties.

57300 replied:
Sorry, `zephyr_topological_sort()` needs to stay generic in order to support groups: https://github.com/zephyrproject-rtos/zephyr/pull/58174

Please have a look at how I extended the `sysbuild_images_order()` function in https://github.com/zephyrproject-rtos/zephyr/pull/58174/commits/314fcf8b3fcb4356b78543d9fd75d11020b7a4cf. There, I construct a single dependency graph based on both image-wise and group-wise dependencies. The target-based sorting you wrote would be applicable for image-wise dependencies only.

Thanks for the effort, though... I should have replied sooner.

tejlmand replied:
> Sorry, `zephyr_topological_sort()` needs to stay generic in order to support groups: #58174

Groups are handled outside of `topological_sort()` in a `foreach()` here:
https://github.com/zephyrproject-rtos/zephyr/blob/2b791a1eb2dc489001b6bc244504dadd726e750f/share/sysbuild/cmake/modules/sysbuild_extensions.cmake#L680
https://github.com/zephyrproject-rtos/zephyr/blob/2b791a1eb2dc489001b6bc244504dadd726e750f/share/sysbuild/cmake/modules/sysbuild_extensions.cmake#L816

so I don't see the relevance in relation to the groups PR.

Instead, the use of `,`-separated strings just makes it more complex to use the topological sort function, as every caller must now generate a list of such strings.

A core part of CMake are targets and handling of those, so sticking to targets and properties will remove need for construction of `,`-separated strings.

And groups are already defined as targets: https://github.com/zephyrproject-rtos/zephyr/blob/2b791a1eb2dc489001b6bc244504dadd726e750f/share/sysbuild/cmake/modules/sysbuild_extensions.cmake#LL433C5-L433C22

57300 replied:
I decided to copy your target-based toposort and credit you as the author, if you don't mind. Thanks again for the effort!

tejlmand commented on share/sysbuild/CMakeLists.txt:
```diff
@@ -94,33 +90,35 @@ add_subdirectory(bootloader)
 
-sysbuild_module_call(PRE_CMAKE MODULES ${SYSBUILD_MODULE_NAMES} IMAGES ${IMAGES})
-foreach(image ${IMAGES})
+sysbuild_images_order(IMAGES_CONFIGURATION_ORDER CONFIGURE IMAGES ${IMAGES})
+sysbuild_images_order(IMAGES_FLASHING_ORDER FLASH IMAGES ${IMAGES})
```
> Sure, why not?

because IMAGES_CONFIGURATION_ORDER doesn't seem to be used as a global variable, it is being (correctly) passed as argument to the `sysbuild_module_call(PRE_CMAKE ...)` function.
Secondly, the IMAGES_CONFIGURATION_ORDER  is correctly used for the `foreach()` loop, but we don't promise a specific image order being passed to modules, ref: https://docs.zephyrproject.org/latest/develop/modules.html#sysbuild-modules-hooks

Meaning users should not start to depend on a given order.
Unless there is a strong use-case for guaranteeing a specific image order for `sysbuild_module_call(...)` calls, I would prefer we don't start to give such gurantees as it might put restrictions on restructuring sysbuild code.

Also correct that IMAGES_FLASHING_ORDER  is used by domains.cmake, but why doesn't domains.cmake do the sorting itself ?

So far I see no specific argument as to why `sysbuild_module_call(PRE_DOMAINS ...)` needs the images to be ordered.

tejlmand replied:
> For simplicity, I wanted to have a single cut-off point after which it's no longer valid to add more images or update their properties.

Understandable, but that's not even the reality today.
User's can adjust any image property / cache setting until CMake configure has been invoked.
So I don't see why the image dependency should have special treatment, as it's just one setting among others.

Regarding the `PRE_DOMAINS` you may argue that we don't offer any special handling / properties dedicated to this.
Today the PRE/POST_DOMAINS hooks are provided to allow downstream users to perform custom tasks.
A use-case is merging / post processing of images before they are flashed.

However, if there is no technical reasons why users should be prevented from adjusting properties available, then I don't think we should introduce such limitation.

> because my impression was that they were mainly added to support downstream functionality like signing and such.

That is one use-case, but we don't put restrictions as to what users want to do in those hooks.

> Another concern I have is that hooks are only available for modules and not for individual sysbuild.cmake files. 

Correct, because the sysbuild.cmake serves a different purpose.
There are two kind of sysbuild.cmake files.
- Board specific
- Image specific

For the board specific one, it allows a board to adjust CMake variables, add board required images, etc.
You seem to be considering the image part, and one such image could be a board which needs to have a specific hex / bin flashed, one such example could be the Raspberry Pi Pico, ref: https://github.com/zephyrproject-rtos/zephyr/blob/3104ad0e393cc4df756b2d12aa15c0b3c9be3960/modules/hal_rpi_pico/CMakeLists.txt#L13-L15

If a board specific image is depending on user controlled images, then it's probably not so board specific, but should instead be a generally defined image which is described in Kconfig.

Similar with a sample specific image. If using `<sample>/sysbuild.cmake` to add extra images, then it should be core images to the sample. If they are general purpose samples they should be available in the modules CMakeLists txt files.
And with relevant Kconfig settings and dependencies.

So the hooks are purposely available to Zephyr modules only to help guide users into a proper CMake / Kconfig structure.
The `sysbuild.cmake` files are more specific fine-tuning / specific handling as just described.

Feel free to extend / improve our documentation, as it could be valuable to make the difference / intentions clear for the `sysbuild.cmake` files.

> Assuming that Zephyr itself is not a module, it cannot have its own hooks,

Zephyr itself is not expected to be using those hooks.
If sysbuild has special needs, then it should create a clear step, for example a built-in signing step between `CMake` and `domains`, and then one such step can provide dedicated pre/post hooks.

This is also why there is both a `PRE_DOMAINS` and `POST_CMAKE` even though POST_DOMAINS is called right after POST_CMAKE, cause that ensure users can place their processing at correct location even if more steps are added later.

>  we want that information to be more self-contained, then one option would be to support adding EXTRA_ZEPHYR_MODULES in sysbuild context, but that would require more work.

That is already supported in sysbuild, unless you refer to the NCS misuse of having samples adding themselves as a Zephyr module (which was never the intention of that setting), but guess that's another topic.

57300 replied:
Thanks for the detailed reply. What you wrote gave me more clarity on how this new feature should fit into sysbuild. Much of this info is indeed worth capturing in the documentation; I wouldn't mind looking into it on a separate occasion.

I've made it possible to add `CONFIGURE` deps in `PRE_CMAKE` and `FLASH` deps in `PRE_DOMAINS`.

> If using `<sample>/sysbuild.cmake` to add extra images, then it should be core images to the sample. If they are general purpose samples they should be available in the modules CMakeLists txt files. And with relevant Kconfig settings and dependencies.

That works for me, because those general purpose samples from modules are added early on (and in a well-defined order, as you mentioned elsewhere), so adding dependencies to them should be easy.

The only leftover concern I had was: what if a general purpose sample had its own core image added via `sysbuild.cmake`? Then, adding dependencies to that second image would be trickier, since it wouldn't be added at the same time as the first image. For this reason, I chose to add another patch which adjusts the image processing order.

tejlmand commented on cmake/modules/extensions.cmake:
```diff
@@ -2814,6 +2814,91 @@ function(target_byproducts)
 endfunction()
 
+# Usage:
+#   zephyr_topological_sort(VERTICES <v_0> [<v_1> ...]
+#                           [EDGES <v_i>,<v_j> ...]
+#                           OUTPUT_RESULT <out-variable>
+#                           [OUTPUT_CYCLE <out-variable>])
```
Instead of trying to make this completely generic then make it dependent on targets.

`add_dependencies()` in CMake already requires targets as arguments, and so far I only see use-cases in CMake where topological sort is performed on target.

Therefore, using targets as input will make API and calling code cleaner, for example:
```
#   zephyr_topological_sort(TARGETS<t0> [<t1> ...]
#                           [PROPERTY_NAME <name>]
#                           OUTPUT_RESULT <out-variable>
#                           [OUTPUT_CYCLE <out-variable>])
#
# PROPERTY_NAME <name>: Name of property on <targets> which contains a list of dependencies used for sorting.
#                       As example, a CMake library property name could be `LINK_DEPENDS`.
```

57300 replied:
Addressed here: https://github.com/zephyrproject-rtos/zephyr/pull/57884#discussion_r1206841170

57300 committed:
cmake: extensions: Add zephyr_topological_sort()

This function performs topological sorting of CMake targets. Its initial
use case in Zephyr will be for implementing sysbuild image dependencies,
i.e., specifying an image order for CMake configuration and flashing.

Sourced from a comment on PR #57884 (anchor: #discussion_r1206807021)

Authored-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>
Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>

nordicjm reviewed (commented):
Seems OK, but there are users already requesting ways of building images without flashing them (including zephyr projects), and this seems to prevent being able to do that anymore, so would be good if there was a flag added that could be used when adding an image which would not configure it for flashing.

57300 commented:
> Seems OK, but there are users already requesting ways of building images without flashing them (including zephyr projects), and this seems to prevent being able to do that anymore, so would be good if there was a flag added that could be used when adding an image which would not configure it for flashing.

@nordicjm I assume you're referring to the removal of `IMAGES` as a hindrance here. That was also brought up in this thread: https://github.com/zephyrproject-rtos/zephyr/pull/57884#discussion_r1201620427. Now that this PR is rebased on top of 93b1fb27133dae0e79a7fab283b67bab41b7114d, it shouldn't be a problem.

By the way, I also added an optimization for build-only images in the latest commit.

tejlmand reviewed (changes requested):
Thanks for this huge improvement.

The general impression is very positive, but a couple of must fixes, mostly to avoid surprises.

Regarding the looping itself and comments on `i` / `j`, `IMAGES` local var and `include()`, then I believe we could improve the solution even further wrt. loop counters and robustness with the following approach.
1. Have a custom property on each image identifying if it has already had its `<image_source_dir>/sysbuild.cmake` file processed or not.
2. Whenever a call is made to `ExternalZephyrProject_Add()` , then not only do we add to the global IMAGES property, we also add the image to the IMAGE property of the directory. To allow nested calls of `sysbuild_add_directory()` combined with regular `add_directory()` then we define `SYSBUILD_CURRENT_SOURCE_DIR` on the scope.
   This will allow fetching the IMAGES property and thus obtain all images added in SYSBUILD_CURRENT_SOURCE_DIR. Meaning all images in the current sysbuild_add_subdirectory() and loop those in a `foreach()` without the need for `i` and `j` counters. This should result in more robust code.

Note, such improvement is not a requirement for this PR to be approved, but can come as followup. Feel free to make direct ping for extra details.

57300 committed:
sysbuild: Make the image processing order well-defined

Adjust the order in which image-specific `sysbuild.cmake` files are
iteratively included by sysbuild.

This is motivated by the introduction of `sysbuild_add_dependencies()`.
In the following example:

  sysbuild_add_dependencies(CONFIGURE my_sample sample_a sample_b)

the `my_sample` image is expected to be added before this function is
called. Success depends not only on the placement of the call, but on
the order in which new images are added, which itself is influenced by
the order in which `sysbuild.cmake` files are included. This last order
can be tweaked to make the "dependencies" feature more user-friendly.

This is done by rewriting the internal `sysbuild.cmake` processing loop
into a new, general purpose function - `sysbuild_add_subdirectory()` -
which is a wrapper for `add_subdirectory(<source_dir>)` that recursively
includes `sysbuild.cmake` files for all images found in `<source_dir>`.

With the new function, all images that are expected to be found in a
given `<source_dir>` are guaranteed to be added around the same time.
This wasn't the case with the old processing loop, because the image-
specific `sysbuild.cmake` files (where "sub-images" could be defined)
were left to be processed at the very end.

Below is the initial order in which sysbuild will add all images.
Note: the order of Zephyr modules (from 1 to n) is well-defined.

  1. Main application (aka DEFAULT_IMAGE)
  2. MCUboot (optional)
  3. All images added via these directories:
     3.1. <module-1>.sysbuild-cmake
     3.2. <module-2>.sysbuild-cmake
     ...
     3.n. <module-n>.sysbuild-cmake

  4. All images added via these files:
     4.1. ${BOARD_DIR}/sysbuild.cmake
     4.2. ${APP_DIR}/sysbuild.cmake (aka sub-images of DEFAULT_IMAGE)

These images are intended to be sorted for the users' convenience, from
most to least important in the build system, or least to most dependent
on other images for configuration (potentially).

Finally, the use of `sysbuild_add_subdirectory()` requires updating the
directory structure in sysbuild:

  ./images
    - All images should belong here. The `DEFAULT_IMAGE` should be the
      first and only image at the top level, so that it gets added first
      and its sub-images get added last.

  ./images/bootloader
    - Moved from ./bootloader.

  ./images/boards
    - Adds images from the board-specific `sysbuild.cmake` file.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>

57300 commented:
> Regarding the looping itself and comments on `i` / `j`, `IMAGES` local var and `include()`, then I believe we could improve the solution even further wrt. loop counters and robustness with the following approach.
> 
> 1. Have a custom property on each image identifying if it has already had its `<image_source_dir>/sysbuild.cmake` file processed or not.
> 
> 2. Whenever a call is made to `ExternalZephyrProject_Add()` , then not only do we add to the global IMAGES property, we also add the image to the IMAGE property of the directory. To allow nested calls of `sysbuild_add_directory()` combined with regular `add_directory()` then we define `SYSBUILD_CURRENT_SOURCE_DIR` on the scope.
>        This will allow fetching the IMAGES property and thus obtain all images added in SYSBUILD_CURRENT_SOURCE_DIR. Meaning all images in the current sysbuild_add_subdirectory() and loop those in a `foreach()` without the need for `i` and `j` counters. This should result in more robust code.
> 
> 
> Note, such improvement is not a requirement for this PR to be approved, but can come as followup. Feel free to make direct ping for extra details.

Thank you for this proposal @tejlmand. We talked about it more offline, and I chose to implement `sysbuild_add_subdirectory()` right away, because I think it's a great way to tackle the local variable problem that you rightfully pointed out.

The latest revision includes an updated directory structure to accommodate this function.

57300 commented on share/sysbuild/cmake/domains.cmake:
```diff
@@ -2,16 +2,32 @@
 # SPDX-License-Identifier: Apache-2.0
 
-sysbuild_images_order(IMAGES_FLASHING_ORDER FLASH IMAGES ${IMAGES})
+set(IMAGES_BUILD_ONLY "")
```
> Why not fail the `sysbuild_add_dependencies(FLASH ...)` when calling the function if the images is build only ?

We can do that. The outcome will be that if an image is added with `BUILD_ONLY <condition>`, then the user will have to wrap the `sysbuild_add_dependencies(FLASH ...)` call in the same conditional.

I added an error check and dropped this commit.

57300 commented on share/sysbuild/cmake/domains.cmake:
```diff
@@ -2,16 +2,32 @@
+# This ensures that build-only images will be excluded from flashing,
+# and that any flashing dependencies among them will be ignored as well.
+foreach(image ${IMAGES})
+  get_property(image_is_build_only TARGET ${image} PROPERTY BUILD_ONLY)
```
No objections from me, as long as this can be considered for a later PR. We can also discuss how pre-built images should be represented.

tejlmand replied:
it can be considered for later PR.
Reason for raising it here was because the use of the flag in this new code made me notice it.

57300 commented on share/sysbuild/CMakeLists.txt:
```diff
@@ -24,9 +24,60 @@ find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE} COMPONENTS ${zephyr_modules
+# This encapsulates the set of images encountered in a given <path>, which helps
+# with keeping a well-defined image processing order.
+macro(process path)
+  list(LENGTH IMAGES i)
```
Removed.

57300 commented on share/sysbuild/CMakeLists.txt:
```diff
@@ -24,9 +24,60 @@ find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE} COMPONENTS ${zephyr_modules
+
+  get_filename_component(path_absolute ${path} ABSOLUTE)
+  if(IS_DIRECTORY ${path_absolute})
+    set(binary_dir ${ARGN})
```
Removed.

57300 commented on share/sysbuild/CMakeLists.txt:
```diff
@@ -24,9 +24,60 @@ find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE} COMPONENTS ${zephyr_modules
+    set(binary_dir ${ARGN})
+    add_subdirectory(${path} ${binary_dir})
+  else()
+    include(${path} OPTIONAL)
```
Removed.

57300 commented on share/sysbuild/CMakeLists.txt:
```diff
@@ -24,9 +24,60 @@ find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE} COMPONENTS ${zephyr_modules
+    endif()
+
+    ExternalProject_Get_property(${image} SOURCE_DIR)
+    include(${SOURCE_DIR}/sysbuild.cmake OPTIONAL)
```
Removed.

57300 commented on share/sysbuild/CMakeLists.txt:
```diff
@@ -24,9 +24,60 @@ find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE} COMPONENTS ${zephyr_modules
+# and iteratively process all of the images added there, in the outermost scope.
+# This encapsulates the set of images encountered in a given <path>, which helps
+# with keeping a well-defined image processing order.
+macro(process path)
```
Removed. After chatting offline, I went with your revised proposal with `sysbuild_add_subdirectory()` only.

nordicjm commented on samples/subsys/ipc/openamp/sysbuild.cmake:
```diff
@@ -9,9 +9,13 @@ ExternalZephyrProject_Add(
+# Add dependencies so that the remote sample will be built first
 # This is required because some primary cores need information from the
 # remote core's build, such as the output image's LMA
 add_dependencies(openamp openamp_remote)
```
Is this line still needed?

57300 replied:
I guess so, if it's meant to control the order in which those images are built (not configured).

tejlmand commented on share/sysbuild/images/CMakeLists.txt:
```diff
@@ -0,0 +1,36 @@
+sysbuild_add_subdirectory(bootloader)
+
+# Include zephyr modules generated sysbuild CMake file.
+foreach(SYSBUILD_CURRENT_MODULE_NAME ${SYSBUILD_MODULE_NAMES})
+  # Note the second, binary_dir parameter requires the added
+  # subdirectory to have its own, local cmake target(s). If not then
+  # this binary_dir is created but stays empty. Object files land in
+  # the main binary dir instead.
+  # https://cmake.org/pipermail/cmake/2019-June/069547.html
+  zephyr_string(SANITIZE TOUPPER MODULE_NAME_UPPER ${SYSBUILD_CURRENT_MODULE_NAME})
+  if(NOT ${SYSBUILD_${MODULE_NAME_UPPER}_CMAKE_DIR} STREQUAL "")
+    set(SYSBUILD_CURRENT_MODULE_DIR ${SYSBUILD_${MODULE_NAME_UPPER}_MODULE_DIR})
+    set(SYSBUILD_CURRENT_CMAKE_DIR ${SYSBUILD_${MODULE_NAME_UPPER}_CMAKE_DIR})
+    sysbuild_add_subdirectory(${SYSBUILD_CURRENT_CMAKE_DIR}
+                              ${CMAKE_BINARY_DIR}/modules/${SYSBUILD_CURRENT_MODULE_NAME})
+  endif()
+endforeach()
```
I understand why this is moved here, as today's main benefit of Sysbuild modules is to include extra images to the build.

But as sysbuild evolves in future, then a Sysbuild module may do more / others things than image adding, in which case having this processing placed under `images` may seem unlogical.

I do believe we should be careful to move this under the `images` folder.

57300 replied:
Well, the only reason this is moved here is for the processing order. But you have a good point about placing this under `images` and what this might communicate. I don't mean to imply that the sole purpose of sysbuild modules (or `sysbuild.cmake` files for that matter) is to add more images.

I see this as a naming issue. The directory doesn't need to be named `images`, but I don't know what else to call it. It could be more along the lines of `main`, `root`, or `core`, but those don't sit right with me.

That directory only exists to be the outermost one included with `sysbuild_add_subdirectory()` - other than that, it's completely artificial. Once it's renamed, I think the sysbuild modules should stay here.

tejlmand replied:
> That directory only exists to be the outermost one included with sysbuild_add_subdirectory()

if that's the reason then we could also consider to just create `app` and `boards` folders with a CMakeLists.txt file in each, and then at the top-level `sysbuild/CMakeLists.txt` have:
```
# sysbuild/CMakeLists.txt
...
sysbuild_add_subdirectory(app)
sysbuild_add_subdirectory(bootloader)

# sysbuild modules loop
...
foreach(SYSBUILD_CURRENT_MODULE_NAME ${SYSBUILD_MODULE_NAMES})
  ...
  sysbuild_add_subdirectory(${SYSBUILD_CURRENT_CMAKE_DIR} ...)
endforeach()
...

sysbuild_add_subdirectory(boards)
```

and then remove the `images` folder.

But regardless of whether we do that, or find a better name for the `images` folder, then that is a simple refactoring and can be done after this PR has been merged.

Thanks for sharing the intentions behind the `images` folder.
As we are clear on the purpose of the folder and the naming issues, then let's get this PR, as the functionality is more important than the small refactoring, which can always be done.

tejlmand reviewed (commented):
Thanks for this huge improvement 🎉 

I understand the location of sysbuild modules processing code, but concerned that in future this location might be illogical.
Would like a reply to my comment on this before final approval, as moving this code later may have an impacted / changed behavior on Sysbuild module's processing order.

Other comments to this PR may be addressed later, as addressing those bring no restrictions to use of sysbuild.

57300 commented on share/sysbuild/cmake/modules/sysbuild_extensions.cmake:
```diff
@@ -566,6 +583,45 @@ function(set_config_string image setting value)
+# This continues until all expected images have been added, before returning.
+#
+function(sysbuild_add_subdirectory source_dir)
+  if(ARGC GREATER 2)
+    message(FATAL_ERROR
+      "sysbuild_add_subdirectory(...) called with incorrect number of arguments"
+      " (expected at most 2, got ${ARGC})"
+    )
+  endif()
```
Will remove. I like your other suggestions for this function.

tejlmand reviewed (approved):
lgtm 🎉
